server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSQueryOptionsCheckUtility.java

```diff
@@ -0,0 +1,105 @@
+            new HashMap<Integer, ErrorSeverity>() {
+                {
+                    put(CICSLexer.QUERY, ErrorSeverity.ERROR);
+                    put(CICSLexer.CHANNEL, ErrorSeverity.ERROR);
+                    put(CICSLexer.CONTAINERCNT, ErrorSeverity.ERROR);
+                    put(CICSLexer.RETCODE, ErrorSeverity.ERROR);
+                    put(CICSLexer.COUNTER, ErrorSeverity.ERROR);
+                    put(CICSLexer.DCOUNTER, ErrorSeverity.ERROR);
+                    put(CICSLexer.POOL, ErrorSeverity.ERROR);
+                    put(CICSLexer.VALUE, ErrorSeverity.ERROR);
+                    put(CICSLexer.MINIMUM, ErrorSeverity.ERROR);
+                    put(CICSLexer.MAXIMUM, ErrorSeverity.ERROR);
+                    put(CICSLexer.SECURITY, ErrorSeverity.WARNING);
+                    put(CICSLexer.RESTYPE, ErrorSeverity.ERROR);
+                    put(CICSLexer.RESCLASS, ErrorSeverity.ERROR);
+                    put(CICSLexer.RESIDLENGTH, ErrorSeverity.ERROR);
+                    put(CICSLexer.RESID, ErrorSeverity.ERROR);
+                    put(CICSLexer.LOGMESSAGE, ErrorSeverity.ERROR);
+                    put(CICSLexer.READ, ErrorSeverity.ERROR);
+                    put(CICSLexer.UPDATE, ErrorSeverity.ERROR);
+                    put(CICSLexer.CONTROL, ErrorSeverity.ERROR);
+                    put(CICSLexer.ALTER, ErrorSeverity.ERROR);
+                    put(CICSLexer.NOSUSPEND, ErrorSeverity.WARNING);
+                }
+            };
+
+    public CICSQueryOptionsCheckUtility(
+            DialectProcessingContext context, List<SyntaxError> errors) {
+        super(context, errors, DUPLICATE_CHECK_OPTIONS);
+    }
+
+    /**
+     * Entrypoint to check CICS QUERY rule options
+     *
+     * @param ctx ParserRuleContext subclass containing options
+     * @param <E> A subclass of ParserRuleContext
+     */
+    @Override
+    public <E extends ParserRuleContext> void checkOptions(E ctx) {
+        switch (ctx.getRuleIndex()) {
+            case CICSParser.RULE_cics_query_channel:
+                checkQueryChannel((CICSParser.Cics_query_channelContext) ctx);
+                break;
+            case CICSParser.RULE_cics_query_counter:
+                checkQueryCounter((CICSParser.Cics_query_counterContext) ctx);
+                break;
+            case CICSParser.RULE_cics_query_security:
+                checkQuerySecurity((CICSParser.Cics_query_securityContext) ctx);
+                break;
+            default:
+                break;
+        }
+        checkDuplicates(ctx);
+    }
+
+    private void checkQueryChannel(CICSParser.Cics_query_channelContext ctx) {
+        checkHasMandatoryOptions(ctx.CHANNEL(), ctx, "CHANNEL");
+    }
+
+    private void checkQueryCounter(CICSParser.Cics_query_counterContext ctx) {
+        checkHasExactlyOneOption("COUNTER or DCOUNTER", ctx, ctx.cics_counter_dcounter());
+    }
+
+    private void checkQuerySecurity(CICSParser.Cics_query_securityContext ctx) {
```

Comment:
`RESCLASS` and `RESIDLENGTH` must be specified together.
Also please add tests.